File: frontend/lib/api/developers.ts
import { apiFetch } from "@/lib/api/client";
import type {
  CliResult,
  DeveloperDashboard,
  DeveloperProfile,
  GraphQLResult,
  MarketplaceList,
  MarketplaceReview,
  PlatformWebhook,
  PlatformWebhookLog,
  PluginDetail,
  PluginInstallation,
  PluginLog,
  RestExplorerResult,
  SdkProject,
} from "@/types/developers";

const base = (slug: string) => `/tenants/${slug}/developers`;

export function getDeveloperDashboard(tenantSlug: string) {
  return apiFetch<DeveloperDashboard>(`${base(tenantSlug)}/dashboard`);
}

export function listMarketplace(
  tenantSlug: string,
  params?: { category?: string; plugin_type?: string; search?: string; featured?: boolean },
) {
  const q = new URLSearchParams();
  if (params?.category) q.set("category", params.category);
  if (params?.plugin_type) q.set("plugin_type", params.plugin_type);
  if (params?.search) q.set("search", params.search);
  if (params?.featured) q.set("featured", "true");
  const qs = q.toString();
  return apiFetch<MarketplaceList>(`${base(tenantSlug)}/marketplace${qs ? `?${qs}` : ""}`);
}

export function getPlugin(tenantSlug: string, pluginId: string) {
  return apiFetch<PluginDetail>(`${base(tenantSlug)}/plugins/${pluginId}`);
}

export function listInstalledPlugins(tenantSlug: string) {
  return apiFetch<PluginInstallation[]>(`${base(tenantSlug)}/plugins/installed`);
}

export function installPlugin(tenantSlug: string, pluginSlug: string, settings: Record<string, unknown> = {}) {
  return apiFetch<PluginInstallation>(`${base(tenantSlug)}/plugins/install`, {
    method: "POST",
    body: JSON.stringify({ plugin_slug: pluginSlug, settings }),
  });
}

export function uninstallPlugin(tenantSlug: string, installationId: string) {
  return apiFetch<{ ok: boolean }>(`${base(tenantSlug)}/installations/${installationId}`, {
    method: "DELETE",
  });
}

export function enablePlugin(tenantSlug: string, installationId: string) {
  return apiFetch<PluginInstallation>(`${base(tenantSlug)}/installations/${installationId}/enable`, {
    method: "POST",
  });
}

export function disablePlugin(tenantSlug: string, installationId: string) {
  return apiFetch<PluginInstallation>(`${base(tenantSlug)}/installations/${installationId}/disable`, {
    method: "POST",
  });
}

export function updatePlugin(tenantSlug: string, installationId: string) {
  return apiFetch<PluginInstallation>(`${base(tenantSlug)}/installations/${installationId}/update`, {
    method: "POST",
  });
}

export function listPluginLogs(tenantSlug: string, pluginId?: string) {
  const qs = pluginId ? `?plugin_id=${pluginId}` : "";
  return apiFetch<PluginLog[]>(`${base(tenantSlug)}/logs${qs}`);
}

export function addPluginReview(
  tenantSlug: string,
  pluginId: string,
  payload: { rating: number; title?: string; body?: string },
) {
  return apiFetch<MarketplaceReview>(`${base(tenantSlug)}/plugins/${pluginId}/reviews`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listWebhooks(tenantSlug: string) {
  return apiFetch<PlatformWebhook[]>(`${base(tenantSlug)}/webhooks`);
}

export function createWebhook(
  tenantSlug: string,
  payload: { name: string; url: string; events?: string[]; retry_limit?: number },
) {
  return apiFetch<PlatformWebhook>(`${base(tenantSlug)}/webhooks`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function testWebhook(tenantSlug: string, webhookId: string) {
  return apiFetch<PlatformWebhookLog>(`${base(tenantSlug)}/webhooks/${webhookId}/test`, {
    method: "POST",
  });
}

export function deleteWebhook(tenantSlug: string, webhookId: string) {
  return apiFetch<{ ok: boolean }>(`${base(tenantSlug)}/webhooks/${webhookId}`, { method: "DELETE" });
}

export function listWebhookLogs(tenantSlug: string, webhookId?: string) {
  const qs = webhookId ? `?webhook_id=${webhookId}` : "";
  return apiFetch<PlatformWebhookLog[]>(`${base(tenantSlug)}/webhook-logs${qs}`);
}

export function retryWebhookLog(tenantSlug: string, logId: string) {
  return apiFetch<PlatformWebhookLog>(`${base(tenantSlug)}/webhook-logs/${logId}/retry`, {
    method: "POST",
  });
}

export function listSdkProjects(tenantSlug: string) {
  return apiFetch<SdkProject[]>(`${base(tenantSlug)}/sdk/projects`);
}

export function createSdkProject(
  tenantSlug: string,
  payload: { name: string; project_type?: string; description?: string },
) {
  return apiFetch<SdkProject>(`${base(tenantSlug)}/sdk/projects`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function runApiExplorer(
  tenantSlug: string,
  payload: { method: string; path: string; body?: Record<string, unknown> | null },
) {
  return apiFetch<RestExplorerResult>(`${base(tenantSlug)}/api-explorer`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getGraphqlSchema(tenantSlug: string) {
  return apiFetch<{ sdl: string }>(`${base(tenantSlug)}/graphql/schema`);
}

export function runGraphql(tenantSlug: string, query: string, variables: Record<string, unknown> = {}) {
  return apiFetch<GraphQLResult>(`${base(tenantSlug)}/graphql`, {
    method: "POST",
    body: JSON.stringify({ query, variables }),
  });
}

export function runCli(
  tenantSlug: string,
  payload: { action: string; name?: string; plugin_type?: string; version?: string },
) {
  return apiFetch<CliResult>(`${base(tenantSlug)}/cli`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function upsertDeveloperProfile(
  tenantSlug: string,
  payload: { display_name: string; email?: string; website?: string; bio?: string },
) {
  return apiFetch<DeveloperProfile>(`${base(tenantSlug)}/profile`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

File: backend/app/api/v1/developers.py
"""Developer Platform & Plugin SDK API (Phase 18)."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.deps import TenantContext, require_permission
from app.db.session import get_db
from app.schemas.developer import (
    CliActionRequest,
    CliActionResponse,
    DeveloperDashboardResponse,
    DeveloperResponse,
    DeveloperUpsert,
    GraphQLRequest,
    GraphQLResponse,
    MarketplaceListResponse,
    MarketplaceReviewCreate,
    MarketplaceReviewResponse,
    PlatformWebhookCreate,
    PlatformWebhookLogResponse,
    PlatformWebhookResponse,
    PlatformWebhookUpdate,
    PluginDetail,
    PluginInstallRequest,
    PluginInstallationResponse,
    PluginLogResponse,
    PluginSettingsUpdate,
    RestExplorerRequest,
    RestExplorerResponse,
    SdkProjectCreate,
    SdkProjectResponse,
)
from app.services.developer_service import DeveloperPlatformService

router = APIRouter(prefix="/tenants/{slug}/developers", tags=["developers"])


@router.get("/dashboard", response_model=DeveloperDashboardResponse)
def get_dashboard(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> DeveloperDashboardResponse:
    return DeveloperPlatformService(db).get_dashboard(ctx)


@router.get("/marketplace", response_model=MarketplaceListResponse)
def list_marketplace(
    category: str | None = None,
    plugin_type: str | None = None,
    search: str | None = None,
    featured: bool = False,
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> MarketplaceListResponse:
    return DeveloperPlatformService(db).list_marketplace(
        ctx, category=category, plugin_type=plugin_type, search=search, featured=featured
    )


@router.get("/plugins/installed", response_model=list[PluginInstallationResponse])
def list_installed(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> list[PluginInstallationResponse]:
    return DeveloperPlatformService(db).list_installed(ctx)


@router.post("/plugins/install", response_model=PluginInstallationResponse, status_code=status.HTTP_201_CREATED)
def install_plugin(
    payload: PluginInstallRequest,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PluginInstallationResponse:
    return DeveloperPlatformService(db).install_plugin(ctx, payload)


@router.get("/plugins/{plugin_id}", response_model=PluginDetail)
def get_plugin(
    plugin_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> PluginDetail:
    return DeveloperPlatformService(db).get_plugin(ctx, plugin_id)


@router.post("/plugins/{plugin_id}/reviews", response_model=MarketplaceReviewResponse, status_code=status.HTTP_201_CREATED)
def add_review(
    plugin_id: UUID,
    payload: MarketplaceReviewCreate,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> MarketplaceReviewResponse:
    return DeveloperPlatformService(db).add_review(ctx, plugin_id, payload)


@router.delete("/installations/{installation_id}")
def uninstall_plugin(
    installation_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> dict:
    return DeveloperPlatformService(db).uninstall_plugin(ctx, installation_id)


@router.post("/installations/{installation_id}/enable", response_model=PluginInstallationResponse)
def enable_plugin(
    installation_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PluginInstallationResponse:
    return DeveloperPlatformService(db).set_plugin_enabled(ctx, installation_id, True)


@router.post("/installations/{installation_id}/disable", response_model=PluginInstallationResponse)
def disable_plugin(
    installation_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PluginInstallationResponse:
    return DeveloperPlatformService(db).set_plugin_enabled(ctx, installation_id, False)


@router.post("/installations/{installation_id}/update", response_model=PluginInstallationResponse)
def update_plugin(
    installation_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PluginInstallationResponse:
    return DeveloperPlatformService(db).update_plugin_version(ctx, installation_id)


@router.patch("/installations/{installation_id}/settings", response_model=PluginInstallationResponse)
def update_settings(
    installation_id: UUID,
    payload: PluginSettingsUpdate,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PluginInstallationResponse:
    return DeveloperPlatformService(db).update_plugin_settings(ctx, installation_id, payload)


@router.get("/logs", response_model=list[PluginLogResponse])
def list_logs(
    plugin_id: UUID | None = None,
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> list[PluginLogResponse]:
    return DeveloperPlatformService(db).list_plugin_logs(ctx, plugin_id)


@router.get("/profile", response_model=DeveloperResponse)
def get_profile(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> DeveloperResponse:
    return DeveloperResponse.model_validate(DeveloperPlatformService(db).ensure_developer_profile(ctx))


@router.put("/profile", response_model=DeveloperResponse)
def upsert_profile(
    payload: DeveloperUpsert,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> DeveloperResponse:
    return DeveloperPlatformService(db).upsert_developer(ctx, payload)


@router.get("/sdk/projects", response_model=list[SdkProjectResponse])
def list_sdk_projects(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> list[SdkProjectResponse]:
    return DeveloperPlatformService(db).list_sdk_projects(ctx)


@router.post("/sdk/projects", response_model=SdkProjectResponse, status_code=status.HTTP_201_CREATED)
def create_sdk_project(
    payload: SdkProjectCreate,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> SdkProjectResponse:
    return DeveloperPlatformService(db).create_sdk_project(ctx, payload)


@router.get("/webhooks", response_model=list[PlatformWebhookResponse])
def list_webhooks(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> list[PlatformWebhookResponse]:
    return DeveloperPlatformService(db).list_webhooks(ctx)


@router.post("/webhooks", response_model=PlatformWebhookResponse, status_code=status.HTTP_201_CREATED)
def create_webhook(
    payload: PlatformWebhookCreate,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PlatformWebhookResponse:
    return DeveloperPlatformService(db).create_webhook(ctx, payload)


@router.patch("/webhooks/{webhook_id}", response_model=PlatformWebhookResponse)
def update_webhook(
    webhook_id: UUID,
    payload: PlatformWebhookUpdate,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PlatformWebhookResponse:
    return DeveloperPlatformService(db).update_webhook(ctx, webhook_id, payload)


@router.delete("/webhooks/{webhook_id}")
def delete_webhook(
    webhook_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> dict:
    return DeveloperPlatformService(db).delete_webhook(ctx, webhook_id)


@router.post("/webhooks/{webhook_id}/test", response_model=PlatformWebhookLogResponse)
def test_webhook(
    webhook_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PlatformWebhookLogResponse:
    return DeveloperPlatformService(db).test_webhook(ctx, webhook_id)


@router.post("/webhook-logs/{log_id}/retry", response_model=PlatformWebhookLogResponse)
def retry_webhook_log(
    log_id: UUID,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> PlatformWebhookLogResponse:
    return DeveloperPlatformService(db).retry_webhook_log(ctx, log_id)


@router.get("/webhook-logs", response_model=list[PlatformWebhookLogResponse])
def list_webhook_logs(
    webhook_id: UUID | None = Query(default=None),
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> list[PlatformWebhookLogResponse]:
    return DeveloperPlatformService(db).list_webhook_logs(ctx, webhook_id)


@router.post("/api-explorer", response_model=RestExplorerResponse)
def api_explorer(
    payload: RestExplorerRequest,
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> RestExplorerResponse:
    return DeveloperPlatformService(db).rest_explorer(ctx, payload)


@router.get("/graphql/schema")
def graphql_schema(
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> dict:
    return DeveloperPlatformService(db).graphql_sdl()


@router.post("/graphql", response_model=GraphQLResponse)
def graphql_query(
    payload: GraphQLRequest,
    ctx: TenantContext = Depends(require_permission("developer:read")),
    db: Session = Depends(get_db),
) -> GraphQLResponse:
    return DeveloperPlatformService(db).graphql(ctx, payload)


@router.post("/cli", response_model=CliActionResponse)
def run_cli(
    payload: CliActionRequest,
    ctx: TenantContext = Depends(require_permission("developer:write")),
    db: Session = Depends(get_db),
) -> CliActionResponse:
    return DeveloperPlatformService(db).run_cli(ctx, payload)
